File: py/48.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """   
        head = 0
        tail = len(matrix) - 1

        while head < tail:
            matrix[head], matrix[tail] = matrix[tail], matrix[head]
            head += 1
            tail -= 1
    
        for i in range(len(matrix)):
            for j in range(i + 1, len(matrix[0])):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

File: java/48.java
class Solution {
    public void rotate(int[][] matrix) {
        int head = 0;
        int tail = matrix.length - 1;

        while (head < tail) {
            for (int i = 0; i < matrix.length; i++){
                int tmp;
                tmp = matrix[head][i];
                matrix[head][i] = matrix[tail][i];
                matrix[tail][i] = tmp;
            }
            head ++;
            tail --;
        }

        for (int i = 0; i < matrix.length; i++) {
            for (int j = i + 1; j < matrix.length; j++) {
                int tmp;
                tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }        
    }
}